File: frontend/src/services/expensePredictionService.js
// frontend/src/services/expensePredictionService.js

import {
  escucharHistorialFinanciero,
  recopilarHistorialFinanciero,
} from './financialHistoryService';

export const ESTADOS_PREDICCION_GASTOS = {
  GENERADA: 'generada',
  DATOS_INSUFICIENTES: 'datos_insuficientes',
  ERROR: 'error',
};

const NOMBRE_MODELO = 'modelo_predictivo_gastos_combinado';
const VERSION_MODELO = '1.0';

const redondear = (valor, decimales = 2) => {
  const numero = Number(valor) || 0;
  const factor = 10 ** decimales;

  return Math.round((numero + Number.EPSILON) * factor) / factor;
};

const obtenerNombreMesDesdeFecha = (fecha) => {
  return fecha.toLocaleDateString('es-CO', {
    year: 'numeric',
    month: 'long',
  });
};

const obtenerClaveMesDesdeFecha = (fecha) => {
  const year = fecha.getFullYear();
  const month = String(fecha.getMonth() + 1).padStart(2, '0');

  return `${year}-${month}`;
};

const convertirMesAFecha = (mes) => {
  if (!mes || mes === 'sin-fecha') return null;

  const [year, month] = mes.split('-');

  const fecha = new Date(Number(year), Number(month) - 1, 1);

  if (Number.isNaN(fecha.getTime())) {
    return null;
  }

  return fecha;
};

const obtenerSerieGastosMensuales = (historialMensual = []) => {
  return historialMensual
    .filter((item) => item.mes && item.mes !== 'sin-fecha')
    .map((item) => ({
      mes: item.mes,
      nombreMes: item.nombreMes,
      gastos: Number(item.gastos) || 0,
      totalMovimientos: Number(item.totalMovimientos) || 0,
    }))
    .sort((a, b) => a.mes.localeCompare(b.mes));
};

const calcularPromedio = (valores) => {
  if (!valores.length) return 0;

  const total = valores.reduce((suma, valor) => suma + valor, 0);

  return redondear(total / valores.length);
};

const calcularDesviacionEstandar = (valores) => {
  if (!valores.length) return 0;

  const promedio = calcularPromedio(valores);

  const varianza =
    valores.reduce((suma, valor) => {
      return suma + (valor - promedio) ** 2;
    }, 0) / valores.length;

  return redondear(Math.sqrt(varianza));
};

const calcularPromedioPonderadoReciente = (valores, limite = 3) => {
  if (!valores.length) return 0;

  const ultimosValores = valores.slice(-limite);

  let sumaPesos = 0;
  let sumaPonderada = 0;

  ultimosValores.forEach((valor, index) => {
    const peso = index + 1;

    sumaPesos += peso;
    sumaPonderada += valor * peso;
  });

  return sumaPesos ? redondear(sumaPonderada / sumaPesos) : 0;
};

const calcularRegresionLineal = (valores) => {
  const n = valores.length;

  if (n < 2) {
    return {
      pendiente: 0,
      intercepto: valores[0] || 0,
      r2: 0,
      descripcion: 'No hay suficientes meses para calcular tendencia lineal.',
    };
  }

  const puntos = valores.map((valor, index) => ({
    x: index,
    y: valor,
  }));

  const sumaX = puntos.reduce((suma, punto) => suma + punto.x, 0);
  const sumaY = puntos.reduce((suma, punto) => suma + punto.y, 0);
  const sumaXY = puntos.reduce((suma, punto) => suma + punto.x * punto.y, 0);
  const sumaX2 = puntos.reduce((suma, punto) => suma + punto.x ** 2, 0);

  const denominador = n * sumaX2 - sumaX ** 2;

  const pendiente =
    denominador === 0
      ? 0
      : (n * sumaXY - sumaX * sumaY) / denominador;

  const intercepto = (sumaY - pendiente * sumaX) / n;

  const promedioY = sumaY / n;

  const ssTotal = puntos.reduce(
    (suma, punto) => suma + (punto.y - promedioY) ** 2,
    0
  );

  const ssResiduo = puntos.reduce((suma, punto) => {
    const estimado = intercepto + pendiente * punto.x;

    return suma + (punto.y - estimado) ** 2;
  }, 0);

  const r2 = ssTotal === 0 ? 1 : 1 - ssResiduo / ssTotal;

  return {
    pendiente: redondear(pendiente),
    intercepto: redondear(intercepto),
    r2: redondear(Math.max(0, Math.min(1, r2)), 4),
    descripcion: 'Tendencia calculada mediante regresión lineal simple.',
  };
};

const describirTendencia = ({ pendiente, promedio }) => {
  if (!promedio || promedio <= 0) {
    return {
      tendencia: 'insuficiente',
      descripcion: 'No hay gastos promedio suficientes para describir tendencia.',
    };
  }

  const variacionMensual = redondear((pendiente / promedio) * 100);

  if (variacionMensual > 8) {
    return {
      tendencia: 'aumento',
      variacionMensual,
      descripcion: `Los gastos muestran una tendencia de aumento aproximada de ${variacionMensual}% mensual.`,
    };
  }

  if (variacionMensual < -8) {
    return {
      tendencia: 'disminucion',
      variacionMensual,
      descripcion: `Los gastos muestran una tendencia de disminución aproximada de ${Math.abs(
        variacionMensual
      )}% mensual.`,
    };
  }

  return {
    tendencia: 'estable',
    variacionMensual,
    descripcion: 'Los gastos se mantienen relativamente estables.',
  };
};

const limitarEstimacion = ({ estimacion, promedio }) => {
  if (!Number.isFinite(estimacion) || estimacion < 0) {
    return 0;
  }

  if (!promedio || promedio <= 0) {
    return redondear(estimacion);
  }

  const maximoPermitido = promedio * 2.2;

  return redondear(Math.min(estimacion, maximoPermitido));
};

const calcularRangoPrediccion = ({ gastoEstimado, desviacionEstandar }) => {
  const margenMinimo = gastoEstimado * 0.12;
  const margenEstadistico = desviacionEstandar * 0.7;
  const margen = Math.max(margenMinimo, margenEstadistico);

  return {
    rangoMinimo: redondear(Math.max(0, gastoEstimado - margen)),
    rangoMaximo: redondear(gastoEstimado + margen),
    margenErrorEstimado: redondear(margen),
  };
};

const calcularConfianzaPrediccion = ({
  totalMeses,
  totalMovimientos,
  coeficienteVariacion,
  r2,
}) => {
  let puntaje = 35;

  if (totalMeses >= 6) {
    puntaje += 25;
  } else if (totalMeses >= 3) {
    puntaje += 15;
  } else if (totalMeses > 0) {
    puntaje += 5;
  }

  if (totalMovimientos >= 30) {
    puntaje += 20;
  } else if (totalMovimientos >= 10) {
    puntaje += 12;
  } else if (totalMovimientos > 0) {
    puntaje += 5;
  }

  if (coeficienteVariacion <= 20) {
    puntaje += 20;
  } else if (coeficienteVariacion <= 40) {
    puntaje += 12;
  } else if (coeficienteVariacion <= 60) {
    puntaje += 5;
  } else {
    puntaje -= 5;
  }

  if (r2 >= 0.6) {
    puntaje += 10;
  } else if (r2 >= 0.3) {
    puntaje += 5;
  }

  puntaje = Math.max(0, Math.min(100, Math.round(puntaje)));

  if (puntaje >= 75) {
    return {
      nivel: 'alta',
      puntaje,
      descripcion: 'La predicción tiene buena base histórica.',
    };
  }

  if (puntaje >= 50) {
    return {
      nivel: 'media',
      puntaje,
      descripcion: 'La predicción es útil, pero debe revisarse con cautela.',
    };
  }

  return {
    nivel: 'baja',
    puntaje,
    descripcion: 'La predicción es aproximada y requiere más historial financiero.',
  };
};

const calcularCategoriasEstimadas = ({
  categoriasFrecuentes = [],
  gastoEstimado,
}) => {
  const totalCategorias = categoriasFrecuentes.reduce(
    (suma, categoria) => suma + (Number(categoria.total) || 0),
    0
  );

  if (!totalCategorias || totalCategorias <= 0) {
    return [];
  }

  return categoriasFrecuentes.slice(0, 5).map((categoria) => {
    const porcentaje = redondear(
      ((Number(categoria.total) || 0) / totalCategorias) * 100
    );

    return {
      categoria: categoria.categoria,
      porcentajeHistorico: porcentaje,
      gastoEstimado: redondear((gastoEstimado * porcentaje) / 100),
    };
  });
};

const obtenerFechaUltimoMesHistorico = (serieGastos) => {
  if (!serieGastos.length) {
    return new Date();
  }

  const ultimoMes = serieGastos[serieGastos.length - 1].mes;

  return convertirMesAFecha(ultimoMes) || new Date();
};

const calcularPrediccionesMensuales = ({
  serieGastos,
  mesesPrediccion,
  promedioHistorico,
  promedioPonderado,
  desviacionEstandar,
  regresion,
  categoriasFrecuentes,
}) => {
  const predicciones = [];
  const fechaBase = obtenerFechaUltimoMesHistorico(serieGastos);
  const ultimoGasto = serieGastos[serieGastos.length - 1]?.gastos || promedioHistorico;

  for (let i = 1; i <= mesesPrediccion; i += 1) {
    const fechaPrediccion = new Date(
      fechaBase.getFullYear(),
      fechaBase.getMonth() + i,
      1
    );

    const indiceFuturo = serieGastos.length + i - 1;

    const estimacionTendencia =
      serieGastos.length >= 2
        ? regresion.intercepto + regresion.pendiente * indiceFuturo
        : promedioHistorico;

    const estimacionCombinada =
      serieGastos.length >= 2
        ? estimacionTendencia * 0.5 +
          promedioPonderado * 0.3 +
          promedioHistorico * 0.2
        : promedioHistorico || ultimoGasto;

    const gastoEstimado = limitarEstimacion({
      estimacion: estimacionCombinada,
      promedio: promedioHistorico,
    });

    const rango = calcularRangoPrediccion({
      gastoEstimado,
      desviacionEstandar,
    });

    predicciones.push({
      mes: obtenerClaveMesDesdeFecha(fechaPrediccion),
      nombreMes: obtenerNombreMesDesdeFecha(fechaPrediccion),
      gastoEstimado,
      ...rango,
      categoriasEstimadas: calcularCategoriasEstimadas({
        categoriasFrecuentes,
        gastoEstimado,
      }),
      explicacion:
        'Estimación calculada combinando promedio histórico, comportamiento reciente y tendencia mensual de gastos.',
    });
  }

  return predicciones;
};

const crearRespuestaSinDatos = ({ historialFinanciero, mesesPrediccion }) => {
  return {
    exito: false,
    estado: ESTADOS_PREDICCION_GASTOS.DATOS_INSUFICIENTES,
    modelo: {
      nombre: NOMBRE_MODELO,
      version: VERSION_MODELO,
      metodo: 'No aplicado por falta de datos.',
    },
    periodoHistorico: historialFinanciero?.periodo || null,
    horizontePrediccion: {
      mesesPrediccion,
    },
    prediccionesMensuales: [],
    prediccionTotal: {
      gastoEstimado: 0,
      rangoMinimo: 0,
      rangoMaximo: 0,
    },
    confianza: {
      nivel: 'baja',
      puntaje: 0,
      descripcion: 'No hay historial suficiente para generar predicciones.',
    },
    advertencias: [
      'No hay historial financiero suficiente para generar predicciones.',
      'Las predicciones son aproximaciones y dependen de la calidad de los datos históricos.',
    ],
    actualizadoEn: new Date().toISOString(),
  };
};

export const calcularPrediccionGastosFuturos = ({
  historialFinanciero,
  mesesPrediccion = 3,
} = {}) => {
  if (!historialFinanciero) {
    throw new Error('No se recibió el historial financiero para la predicción.');
  }

  const serieGastos = obtenerSerieGastosMensuales(
    historialFinanciero.historialMensual || []
  );

  if (!serieGastos.length) {
    return crearRespuestaSinDatos({
      historialFinanciero,
      mesesPrediccion,
    });
  }

  const valoresGastos = serieGastos.map((item) => item.gastos);

  const promedioHistorico = calcularPromedio(valoresGastos);
  const promedioPonderado = calcularPromedioPonderadoReciente(valoresGastos);
  const desviacionEstandar = calcularDesviacionEstandar(valoresGastos);

  const coeficienteVariacion =
    promedioHistorico > 0
      ? redondear((desviacionEstandar / promedioHistorico) * 100)
      : 0;

  const regresion = calcularRegresionLineal(valoresGastos);

  const tendenciaDetectada = describirTendencia({
    pendiente: regresion.pendiente,
    promedio: promedioHistorico,
  });

  const prediccionesMensuales = calcularPrediccionesMensuales({
    serieGastos,
    mesesPrediccion,
    promedioHistorico,
    promedioPonderado,
    desviacionEstandar,
    regresion,
    categoriasFrecuentes: historialFinanciero.categoriasFrecuentes || [],
  });

  const gastoEstimadoTotal = redondear(
    prediccionesMensuales.reduce(
      (suma, prediccion) => suma + prediccion.gastoEstimado,
      0
    )
  );

  const rangoMinimoTotal = redondear(
    prediccionesMensuales.reduce(
      (suma, prediccion) => suma + prediccion.rangoMinimo,
      0
    )
  );

  const rangoMaximoTotal = redondear(
    prediccionesMensuales.reduce(
      (suma, prediccion) => suma + prediccion.rangoMaximo,
      0
    )
  );

  const confianza = calcularConfianzaPrediccion({
    totalMeses: serieGastos.length,
    totalMovimientos: historialFinanciero.totalMovimientosRecopilados || 0,
    coeficienteVariacion,
    r2: regresion.r2,
  });

  return {
    exito: true,
    estado: ESTADOS_PREDICCION_GASTOS.GENERADA,
    modelo: {
      nombre: NOMBRE_MODELO,
      version: VERSION_MODELO,
      metodo:
        'Modelo combinado basado en promedio histórico, promedio ponderado reciente y tendencia lineal simple.',
      variablesUsadas: [
        'gastos mensuales históricos',
        'categorías frecuentes',
        'tendencia de gastos',
        'promedio ponderado reciente',
      ],
    },
    periodoHistorico: historialFinanciero.periodo || null,
    horizontePrediccion: {
      mesesPrediccion,
    },
    resumenHistorico: {
      totalMesesAnalizados: serieGastos.length,
      totalMovimientosAnalizados:
        historialFinanciero.totalMovimientosRecopilados || 0,
      gastoPromedioMensual: promedioHistorico,
      gastoPromedioReciente: promedioPonderado,
      desviacionEstandar,
      coeficienteVariacion,
    },
    tendenciaDetectada: {
      ...tendenciaDetectada,
      pendienteMensual: regresion.pendiente,
      r2: regresion.r2,
    },
    prediccionesMensuales,
    prediccionTotal: {
      gastoEstimado: gastoEstimadoTotal,
      rangoMinimo: rangoMinimoTotal,
      rangoMaximo: rangoMaximoTotal,
    },
    confianza,
    advertencias: [
      'Las predicciones son aproximaciones basadas en datos históricos.',
      'Los resultados pueden cambiar si se registran nuevos movimientos financieros.',
      'Este modelo no reemplaza la revisión personal del usuario sobre sus finanzas.',
      ...(historialFinanciero.advertencias || []),
    ],
    actualizadoEn: new Date().toISOString(),
  };
};

export const generarPrediccionGastosUsuario = async ({
  mesesHistorial = 6,
  mesesPrediccion = 3,
} = {}) => {
  const historialFinanciero = await recopilarHistorialFinanciero({
    meses: mesesHistorial,
  });

  return calcularPrediccionGastosFuturos({
    historialFinanciero,
    mesesPrediccion,
  });
};

export const escucharPrediccionGastosFuturos = ({
  mesesHistorial = 6,
  mesesPrediccion = 3,
  onPrediccion,
  onError,
}) => {
  const cancelarEscucha = escucharHistorialFinanciero({
    meses: mesesHistorial,
    onHistorial: (historialFinanciero) => {
      try {
        const prediccion = calcularPrediccionGastosFuturos({
          historialFinanciero,
          mesesPrediccion,
        });

        if (typeof onPrediccion === 'function') {
          onPrediccion(prediccion);
        }
      } catch (error) {
        if (typeof onError === 'function') {
          onError(error);
        } else {
          console.error('Error al calcular predicción de gastos:', error);
        }
      }
    },
    onError,
  });

  return cancelarEscucha;
};

File: backend-api/services/expense_prediction_service.py
import math
from datetime import datetime
from typing import Any, Dict, List

from services.financial_history_service import collect_financial_history

MODEL_NAME = "modelo_predictivo_gastos_backend"
MODEL_VERSION = "1.0"


def round_number(value: Any, decimals: int = 2) -> float:
    try:
        return round(float(value or 0), decimals)
    except (TypeError, ValueError):
        return 0.0


def average(values: List[float]) -> float:
    if not values:
        return 0.0

    return round_number(sum(values) / len(values))


def std_dev(values: List[float]) -> float:
    if not values:
        return 0.0

    avg = average(values)
    variance = sum((value - avg) ** 2 for value in values) / len(values)

    return round_number(math.sqrt(variance))


def weighted_recent_average(values: List[float], limit: int = 3) -> float:
    if not values:
        return 0.0

    latest = values[-limit:]
    total_weight = 0
    weighted_sum = 0

    for index, value in enumerate(latest):
        weight = index + 1
        total_weight += weight
        weighted_sum += value * weight

    return round_number(weighted_sum / total_weight) if total_weight else 0.0


def linear_regression(values: List[float]) -> Dict[str, Any]:
    n = len(values)

    if n < 2:
        return {
            "pendiente": 0,
            "intercepto": values[0] if values else 0,
            "r2": 0,
        }

    x_values = list(range(n))
    sum_x = sum(x_values)
    sum_y = sum(values)
    sum_xy = sum(x * y for x, y in zip(x_values, values))
    sum_x2 = sum(x**2 for x in x_values)

    denominator = n * sum_x2 - sum_x**2

    slope = 0 if denominator == 0 else (n * sum_xy - sum_x * sum_y) / denominator
    intercept = (sum_y - slope * sum_x) / n

    avg_y = sum_y / n
    ss_total = sum((y - avg_y) ** 2 for y in values)
    ss_residue = sum((y - (intercept + slope * x)) ** 2 for x, y in zip(x_values, values))

    r2 = 1 if ss_total == 0 else 1 - (ss_residue / ss_total)

    return {
        "pendiente": round_number(slope),
        "intercepto": round_number(intercept),
        "r2": round_number(max(0, min(1, r2)), 4),
    }


def describe_trend(slope: float, historical_average: float) -> Dict[str, Any]:
    if historical_average <= 0:
        return {
            "tendencia": "insuficiente",
            "variacionMensual": 0,
            "descripcion": "No hay promedio histórico suficiente.",
        }

    monthly_variation = round_number((slope / historical_average) * 100)

    if monthly_variation > 8:
        return {
            "tendencia": "aumento",
            "variacionMensual": monthly_variation,
            "descripcion": f"Los gastos muestran una tendencia de aumento aproximada de {monthly_variation}% mensual.",
        }

    if monthly_variation < -8:
        return {
            "tendencia": "disminucion",
            "variacionMensual": monthly_variation,
            "descripcion": f"Los gastos muestran una tendencia de disminución aproximada de {abs(monthly_variation)}% mensual.",
        }

    return {
        "tendencia": "estable",
        "variacionMensual": monthly_variation,
        "descripcion": "Los gastos se mantienen relativamente estables.",
    }


def calculate_confidence(
    total_months: int,
    total_movements: int,
    coefficient_variation: float,
    r2: float,
) -> Dict[str, Any]:
    score = 35

    if total_months >= 6:
        score += 25
    elif total_months >= 3:
        score += 15
    elif total_months > 0:
        score += 5

    if total_movements >= 30:
        score += 20
    elif total_movements >= 10:
        score += 12
    elif total_movements > 0:
        score += 5

    if coefficient_variation <= 20:
        score += 20
    elif coefficient_variation <= 40:
        score += 12
    elif coefficient_variation <= 60:
        score += 5
    else:
        score -= 5

    if r2 >= 0.6:
        score += 10
    elif r2 >= 0.3:
        score += 5

    score = max(0, min(100, round(score)))

    if score >= 75:
        return {
            "nivel": "alta",
            "puntaje": score,
            "descripcion": "La predicción tiene buena base histórica.",
        }

    if score >= 50:
        return {
            "nivel": "media",
            "puntaje": score,
            "descripcion": "La predicción es útil, pero debe revisarse con cautela.",
        }

    return {
        "nivel": "baja",
        "puntaje": score,
        "descripcion": "La predicción requiere más historial financiero.",
    }


def month_key_to_date(month_key: str) -> datetime:
    return datetime.strptime(month_key, "%Y-%m")


def next_month_key(base_month: str, offset: int) -> str:
    base_date = month_key_to_date(base_month)
    month = base_date.month + offset
    year = base_date.year

    while month > 12:
        month -= 12
        year += 1

    return f"{year}-{month:02d}"


def calculate_category_estimations(
    frequent_categories: List[Dict[str, Any]],
    estimated_expense: float,
) -> List[Dict[str, Any]]:
    total_categories = sum(float(item.get("total", 0) or 0) for item in frequent_categories)

    if total_categories <= 0:
        return []

    result = []

    for category in frequent_categories[:5]:
        percentage = round_number((float(category["total"]) / total_categories) * 100)

        result.append(
            {
                "categoria": category["categoria"],
                "porcentajeHistorico": percentage,
                "gastoEstimado": round_number((estimated_expense * percentage) / 100),
            }
        )

    return result


def calculate_expense_prediction(
    financial_history: Dict[str, Any],
    prediction_months: int = 3,
) -> Dict[str, Any]:
    monthly_history = financial_history.get("historialMensual", [])

    monthly_expenses = [
        float(item.get("gastos", 0) or 0)
        for item in monthly_history
        if item.get("mes") and item.get("mes") != "sin-fecha"
    ]

    if not monthly_expenses:
        return {
            "exito": False,
            "estado": "datos_insuficientes",
            "modelo": {
                "nombre": MODEL_NAME,
                "version": MODEL_VERSION,
            },
            "periodoHistorico": financial_history.get("periodo"),
            "horizontePrediccion": {
                "mesesPrediccion": prediction_months,
            },
            "prediccionesMensuales": [],
            "prediccionTotal": {
                "gastoEstimado": 0,
                "rangoMinimo": 0,
                "rangoMaximo": 0,
            },
            "confianza": {
                "nivel": "baja",
                "puntaje": 0,
                "descripcion": "No hay historial suficiente.",
            },
            "advertencias": [
                "No hay historial financiero suficiente para generar predicciones.",
                "Las predicciones son aproximaciones basadas en datos históricos.",
            ],
        }

    historical_average = average(monthly_expenses)
    recent_average = weighted_recent_average(monthly_expenses)
    deviation = std_dev(monthly_expenses)

    coefficient_variation = (
        round_number((deviation / historical_average) * 100)
        if historical_average > 0
        else 0
    )

    regression = linear_regression(monthly_expenses)

    trend = describe_trend(regression["pendiente"], historical_average)

    last_month = monthly_history[-1]["mes"]
    predictions = []

    for index in range(1, prediction_months + 1):
        month = next_month_key(last_month, index)
        future_index = len(monthly_expenses) + index - 1

        trend_estimation = regression["intercepto"] + regression["pendiente"] * future_index

        if len(monthly_expenses) >= 2:
            combined = trend_estimation * 0.5 + recent_average * 0.3 + historical_average * 0.2
        else:
            combined = historical_average

        estimated_expense = max(0, min(combined, historical_average * 2.2 if historical_average else combined))
        estimated_expense = round_number(estimated_expense)

        margin = max(estimated_expense * 0.12, deviation * 0.7)

        predictions.append(
            {
                "mes": month,
                "nombreMes": month,
                "gastoEstimado": estimated_expense,
                "rangoMinimo": round_number(max(0, estimated_expense - margin)),
                "rangoMaximo": round_number(estimated_expense + margin),
                "margenErrorEstimado": round_number(margin),
                "categoriasEstimadas": calculate_category_estimations(
                    financial_history.get("categoriasFrecuentes", []),
                    estimated_expense,
                ),
                "explicacion": "Estimación basada en promedio histórico, promedio reciente y tendencia mensual.",
            }
        )

    total_estimated = round_number(sum(item["gastoEstimado"] for item in predictions))
    total_min = round_number(sum(item["rangoMinimo"] for item in predictions))
    total_max = round_number(sum(item["rangoMaximo"] for item in predictions))

    confidence = calculate_confidence(
        total_months=len(monthly_expenses),
        total_movements=financial_history.get("totalMovimientosRecopilados", 0),
        coefficient_variation=coefficient_variation,
        r2=regression["r2"],
    )

    return {
        "exito": True,
        "estado": "generada",
        "modelo": {
            "nombre": MODEL_NAME,
            "version": MODEL_VERSION,
            "metodo": "Promedio histórico, promedio ponderado reciente y tendencia lineal simple.",
        },
        "periodoHistorico": financial_history.get("periodo"),
        "horizontePrediccion": {
            "mesesPrediccion": prediction_months,
        },
        "resumenHistorico": {
            "totalMesesAnalizados": len(monthly_expenses),
            "totalMovimientosAnalizados": financial_history.get("totalMovimientosRecopilados", 0),
            "gastoPromedioMensual": historical_average,
            "gastoPromedioReciente": recent_average,
            "desviacionEstandar": deviation,
            "coeficienteVariacion": coefficient_variation,
        },
        "tendenciaDetectada": {
            **trend,
            "pendienteMensual": regression["pendiente"],
            "r2": regression["r2"],
        },
        "prediccionesMensuales": predictions,
        "prediccionTotal": {
            "gastoEstimado": total_estimated,
            "rangoMinimo": total_min,
            "rangoMaximo": total_max,
        },
        "confianza": confidence,
        "advertencias": [
            "Las predicciones son aproximaciones basadas en datos históricos.",
            "Los resultados pueden cambiar cuando se registren nuevos movimientos.",
            *financial_history.get("advertencias", []),
        ],
    }


def generate_user_expense_prediction(
    uid: str,
    history_months: int = 6,
    prediction_months: int = 3,
) -> Dict[str, Any]:
    history = collect_financial_history(uid, meses=history_months)

    return calculate_expense_prediction(history, prediction_months)